File: MIDTERM/basetrade/GenPerlLib/search_exec.pl
# \file GenPerlLib/search_exec.pl
#
# \author: (c) Copyright Two Roads Technological Solutions Pvt Ltd 2011
#  Address:
# 	 Suite No 353, Evoma, #14, Bhattarhalli,
# 	 Old Madras Road, Near Garden City College,
# 	 KR Puram, Bangalore 560049, India
# 	 +91 80 4190 3551
#

# This perl module is to be used in scripts like
# my @ADDITIONAL_EXEC_PATHS=();
# my $SIM_STRATEGY_EXEC = SearchExec ( "sim_strategy", @ADDITIONAL_EXEC_PATHS ) ;
# 
# Shown an example in ModelScripts/run_sim_and_summarize.pl

use strict;
use warnings;

my $HOME_DIR=$ENV{'HOME'};
my $REPO="basetrade";
my $LIVE_BIN_DIR=$HOME_DIR."/LiveExec/bin";
my $BASETRADE_BIN_DIR=$HOME_DIR."/cvquant_install/basetrade/bin";
my $GENPERLLIB_DIR=$HOME_DIR."/".$REPO."_install/GenPerlLib";
# my $GENPERLLIB_DIR=$HOME_DIR."/".$REPO."/GenPerlLib";

require "$GENPERLLIB_DIR/exists_with_size.pl"; # ExistsWithSize

sub SearchExec {
    my $exec_name_ = shift;
    my @textvec_ = @_ ;
    my @default_exec_paths_ = () ;
    my @combined_list_exec_paths_ = () ;

    if (defined $ENV{'DEPS_INSTALL'}){
        my $JENKINS_BIN_DIR = $ENV{'DEPS_INSTALL'}."/basetrade/bin";
        push ( @combined_list_exec_paths_, $JENKINS_BIN_DIR);
    }
    push ( @default_exec_paths_, $LIVE_BIN_DIR );
    push ( @default_exec_paths_, $BASETRADE_BIN_DIR );
    push @combined_list_exec_paths_, @textvec_ ; # paths given by userR
    push @combined_list_exec_paths_, @default_exec_paths_ ; # paths I had in defaults
    # Search default_paths after the given list
    
    my $found_exec_ = "";
    foreach my $dirpath_ (@combined_list_exec_paths_)
      {
	# For this path, see what the exec path would be if it were to exist
	my $searching_path_ = $dirpath_."/".$exec_name_ ;
	
	if ( ExistsWithSize ( $searching_path_ ) )
	{ # if I find it then effectively return it
	    # In Perl, for that we need to break out of this loop
	    # after setting the return variable
	    $found_exec_ = $searching_path_;
	    last;
	}
      }

    if ( $found_exec_ eq "" ) {
      print "Did not find $exec_name_ anywhere in search paths\n";
    }
    $found_exec_;
}

1;

File: MIDTERM/basetrade/Tools/search_exec.cpp
/**
   \file Tools/search_exec.cpp

   \author: (c) Copyright Two Roads Technological Solutions Pvt Ltd 2011
   Address:
   Suite No 353, Evoma, #14, Bhattarhalli,
   Old Madras Road, Near Garden City College,
   KR Puram, Bangalore 560049, India
   +91 80 4190 3551
*/

#include <iostream>
#include <vector>
#include <stdlib.h>
#include <string>
#include <dirent.h>

std::string FindExecPath(const std::vector<std::string>& dir_to_search_, const std::string& exec_name_) {
  struct dirent* directory_entry_ = NULL;
  for (unsigned i = 0; i < dir_to_search_.size(); i++) {
    DIR* dir_ = opendir(dir_to_search_[i].c_str());

    while (dir_) {
      directory_entry_ = readdir(dir_);
      if (!directory_entry_) {
        break;
      }
      if (directory_entry_->d_type == DT_DIR) {
        continue;
      }
      if (exec_name_.compare(directory_entry_->d_name) == 0) {
        return dir_to_search_[i] + directory_entry_->d_name;
      }
    }
  }
  return "";
}

int main(int argc, char** argv) {
  if (argc < 2) {
    std::cout << "usage: <exec_basename>\n";
    return 0;
  }

  std::string exec_name_ = std::string(argv[1]);
  std::string home_ = getenv("HOME");
  std::string user_name = getenv("USER");
  std::vector<std::string> dir_to_search_;
  dir_to_search_.clear();
  dir_to_search_.push_back("/home/dvctrader/LiveExec/");
  dir_to_search_.push_back(home_ + "/basetrade_install/");

  std::vector<std::string> sub_dir_name_;
  sub_dir_name_.clear();
  // std::string path = std::string(home) + std::string(user_name);
  if (exec_name_.find(".pl") != std::string::npos) {
    sub_dir_name_.push_back("scripts/");
    sub_dir_name_.push_back("ModelScripts/");
  } else if (exec_name_.find(".sh") != std::string::npos) {
    sub_dir_name_.push_back("scripts/");
    sub_dir_name_.push_back("ModelScripts/");
  } else if (exec_name_.find(".py") != std::string::npos) {
    sub_dir_name_.push_back("scripts/");
    sub_dir_name_.push_back("ModelScripts/");
  } else {
    sub_dir_name_.push_back("bin/");
  }

  std::vector<std::string> sub_dir_to_search_;
  sub_dir_to_search_.clear();

  for (unsigned j = 0; j < dir_to_search_.size(); j++) {
    for (unsigned i = 0; i < sub_dir_name_.size(); i++) {
      sub_dir_to_search_.push_back(dir_to_search_[j] + sub_dir_name_[i]);
    }
  }

  std::cout << FindExecPath(sub_dir_to_search_, exec_name_);

  return 0;
}
